plugins/modules/keycloak_authz_authorization_scope.py

```diff
@@ -0,0 +1,255 @@
+                   choices=['present', 'absent']),
+        name=dict(type='str', required=True),
+        display_name=dict(type='str', required=False),
+        icon_uri=dict(type='str', required=False),
+        client_id=dict(type='str', required=True),
+        realm=dict(type='str', required=True)
+    )
+
+    argument_spec.update(meta_args)
+
+    module = AnsibleModule(argument_spec=argument_spec,
+                           supports_check_mode=False,
+                           required_one_of=(
+                               [['token', 'auth_realm', 'auth_username', 'auth_password']]),
+                           required_together=([['auth_realm', 'auth_username', 'auth_password']]))
+
+    result = dict(changed=False, msg='', end_state={})
+
+    # Obtain access token, initialize API
+    try:
+        connection_header = get_token(module.params)
+    except KeycloakError as e:
+        module.fail_json(msg=str(e))
+
+    kc = KeycloakAPI(module, connection_header)
+
+    # Convenience variables
+    state = module.params.get('state')
+    name = module.params.get('name')
+    display_name = module.params.get('display_name')
+    icon_uri = module.params.get('icon_uri')
+    client_id = module.params.get('client_id')
+    realm = module.params.get('realm')
+
+    # Get the "id" of the client based on the usually more human-readable
+    # "clientId"
+    cid = kc.get_client_id(client_id, realm=realm)
+    if not cid:
+        module.fail_json(msg='Invalid client %s for realm %s' %
+                         (client_id, realm))
+
+    # Get current state of the Authorization Scope using its name as the search
+    # filter. This returns False if it is not found.
+    authz_scope = kc.get_authz_authorization_scope_by_name(
+        name=name, client_id=cid, realm=realm)
+
+    # Generate a JSON payload for Keycloak Admin API. This is needed for
+    # "create" and "update" operations.
+    payload = {}
+    payload['name'] = name
+    payload['displayName'] = display_name
+    payload['iconUri'] = icon_uri
+
+    # Add "id" to payload for modify operations
+    if authz_scope:
+        payload['id'] = authz_scope['id']
+
+    # Ensure that undefined (null) optional parameters are presented as empty
+    # strings in the desired state. This makes comparisons with current state
+    # much easier.
+    for k, v in payload.items():
+        if not v:
+            payload[k] = ''
+
+    # Do the above for the current state
+    if authz_scope:
+        for k in ['displayName', 'iconUri']:
+            if k not in authz_scope:
+                authz_scope[k] = ''
+
+    if authz_scope and state == 'present':
+        diff = False
+        for k, v in payload.items():
+            if authz_scope[k] != v:
+                diff = True
+                # At this point we know we have to update the object anyways,
+                # so there's no need to do more work.
+                break
+        if diff:
```

Comment:
Since you already check whether a modification is necessary or not, implementing check mode should be pretty straightforward.

Not implementing check mode should only be done if it requires a lot of extra work, or isn't possible at all with the APIs that are used.

Reply:
I implemented check_mode and diff_mode. I also added integration tests to ensure that those two modes work as expected. Let me know if the changes look good.

Implementing the diff_mode made the difference between snake-case (Ansible module) and camel-case (Keycloak JSON input/output) more visible - I think I'll convert _result['end_state']_ and _result['diff']_ to snake-case in another commit (not today).

Reply:
> Implementing the diff_mode made the difference between snake-case (Ansible module) and camel-case (Keycloak JSON input/output) more visible - I think I'll convert _result['end_state']_ and _result['diff']_ to snake-case in another commit (not today).

I was wrong. The diff looks reasonable as-is:

```
--- snip ---
$ ansible-test integration keycloak_authz_authorization_scope --allow-unsupported -v --diff
TASK [keycloak_authz_authorization_scope : Authorization scope update (remove optional parameters)] ***
--- before                                                                     
+++ after                                                                                                                                                     
@@ -1,6 +1,6 @@                     
 {                                    
-    "displayName": "File delete",                                             
-    "iconUri": "http://localhost/icon.png",                                   
+    "displayName": "",                                                        
+    "iconUri": "",                                                            
     "id": "17b533e1-79f6-4e92-a979-90daf1bfb099",                                                                                                                 "name": "file:delete"                                                                                                                                     }                                                           
--- snip ---
```